src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,191 @@
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index, size + 1);
+        Node<T> tempNode = new Node<>(value);
+        if (index == 0) {
+            addNodeToHead(tempNode);
+        } else {
+            Node<T> currentNode = findNodeByIndex(index - 1);
+            addNodeAfter(currentNode, tempNode);
+            if (index == size) {
+                addNodeToTail(tempNode);
+            }
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, size);
+        return findNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index, size);
+        Node<T> currentNode = findNodeByIndex(index);
+        T oldValue = currentNode.value;
+        currentNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, size);
+        if (isEmpty()) {
+            throw new NoSuchElementException("Cannot remove from an empty list");
+        }
+        T removedValue;
+        if (index == 0) {
+            removedValue = head.value;
+            head = head.next;
+            if (head == null) {
+                tail = null;
+            } else {
+                head.prev = null;
+            }
+        } else {
+            Node<T> currentNode = findNodeByIndex(index);
+            removedValue = currentNode.value;
+            unlink(currentNode);
+        }
+        size--;
+        return removedValue;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        for (int i = 0; i < size; i++) {
+            if (object == null ? current.value == null : object.equals(current.value)) {
+                unlink(current);
+                size--;
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index, int size) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bounds: " + index);
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        Node<T> current;
+        if (index >= size / 2) {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+        } else {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        }
```

Comment:
You can simplify it:
```suggestion
        current = (index >= size / 2) ? tail : head;
        int iterations = (index >= size / 2) ? size - index - 1 : index;
        for (int i = 0; i < iterations; i++) {
             current = (index >= size / 2) ? current.prev : current.next;
        }
```